File: Project_Pytest/page_objects/shopping_cart_page.py
"""
ShoppingCartPage：購物車示範應用的統一頁面物件。

設計原則：
  - 全部以 data-testid 屬性定位元素，與 TestComplete 腳本保持一致。
  - 使用 JavaScript 執行點擊與文字讀取，確保對隱藏元素（hidden attribute /
    display:none）的行為與 TestComplete 的 Click() / contentText 一致。
"""
from selenium.webdriver.common.by import By
from page_objects.base_page import BasePage


class ShoppingCartPage(BasePage):
    """購物車應用統一頁面物件（商品列表 + 購物車 + 結帳 Modal）。"""

    # checkout modal 以 id 定位（無 data-testid）
    _CHECKOUT_MODAL_ID = "checkout-modal"

    # ------------------------------------------------------------------
    # 私有 JS 工具方法
    # ------------------------------------------------------------------

    def _js_text(self, testid: str) -> str:
        """
        以 JS 讀取 data-testid 元素的 textContent（對隱藏元素也有效）。
        對應 TestComplete 的 element.contentText。
        """
        return self.driver.execute_script(
            "var el = document.querySelector('[data-testid=\"" + testid + "\"]');"
            "return el ? el.textContent.trim() : '';"
        )

    def _js_click(self, testid: str) -> None:
        """
        以 JS 點擊 data-testid 元素（對隱藏元素也有效）。
        對應 TestComplete 的 element.Click()。
        """
        self.driver.execute_script(
            "var el = document.querySelector('[data-testid=\"" + testid + "\"]');"
            "if (el) { el.click(); }"
        )

    # ------------------------------------------------------------------
    # 商品列表操作
    # ------------------------------------------------------------------

    def click_add_apple(self) -> None:
        """加入蘋果（add-apple）。"""
        self._js_click("add-apple")

    def click_add_banana(self) -> None:
        """加入香蕉（add-banana）。"""
        self._js_click("add-banana")

    def click_add_milk(self) -> None:
        """加入鮮奶（add-milk）。"""
        self._js_click("add-milk")

    def click_add_product(self, product_id: str) -> None:
        """
        以商品 ID 加入商品（apple / banana / milk）。
        對應 TestComplete 的 control_click_testid("add-{product_id}")。
        """
        self._js_click(f"add-{product_id}")

    # ------------------------------------------------------------------
    # 購物車項目操作（動態元素，依商品 ID）
    # ------------------------------------------------------------------

    def click_qty_plus(self, product_id: str) -> None:
        """點擊 +（增加數量）按鈕。對應 TestComplete qty-plus-{product_id}。"""
        self._js_click(f"qty-plus-{product_id}")

    def click_qty_minus(self, product_id: str) -> None:
        """點擊 -（減少數量）按鈕。對應 TestComplete qty-minus-{product_id}。"""
        self._js_click(f"qty-minus-{product_id}")

    def click_remove(self, product_id: str) -> None:
        """點擊移除按鈕。對應 TestComplete remove-{product_id}。"""
        self._js_click(f"remove-{product_id}")

    def get_qty_value(self, product_id: str) -> str:
        """讀取商品數量顯示值。對應 TestComplete qty-value-{product_id}。"""
        return self._js_text(f"qty-value-{product_id}")

    # ------------------------------------------------------------------
    # 購物車摘要
    # ------------------------------------------------------------------

    def get_cart_count(self) -> str:
        """購物車件數文字，例如 '1 件'。對應 TestComplete cart-count。"""
        return self._js_text("cart-count")

    def get_cart_total(self) -> str:
        """購物車總計文字，例如 'NT$ 30'。對應 TestComplete cart-total。"""
        return self._js_text("cart-total")

    def get_cart_empty_message(self) -> str:
        """空購物車提示文字。對應 TestComplete cart-empty。"""
        return self._js_text("cart-empty")

    # ------------------------------------------------------------------
    # 購物車動作按鈕（在購物車空時這些按鈕的父容器有 hidden 屬性）
    # ------------------------------------------------------------------

    def click_clear_cart(self) -> None:
        """清空購物車按鈕（購物車空時按鈕隱藏，仍以 JS 觸發）。"""
        self._js_click("clear-cart")

    def click_checkout(self) -> None:
        """結帳按鈕（購物車空時按鈕隱藏，仍以 JS 觸發）。"""
        self._js_click("checkout")

    def reset_cart(self) -> None:
        """
        若購物車非空則點清空，確保各測試案例獨立執行。
        對應 TestComplete control_reset_cart()。
        """
        if self.get_cart_count() != "0 件":
            self.click_clear_cart()

    # ------------------------------------------------------------------
    # 結帳 Modal
    # ------------------------------------------------------------------

    def is_checkout_modal_open(self) -> bool:
        """
        回傳 True 若結帳 Modal 目前為開啟狀態。
        以 CSS class 'is-open' 判斷（對應 app.js openCheckoutModal/closeCheckoutModal）。
        """
        modal = self.driver.find_element(By.ID, self._CHECKOUT_MODAL_ID)
        return "is-open" in (modal.get_attribute("class") or "")

    def get_checkout_message(self) -> str:
        """結帳成功訊息文字。對應 TestComplete checkout-message。"""
        return self._js_text("checkout-message")

    def get_checkout_total(self) -> str:
        """結帳總金額文字，例如 '總金額：NT$ 50'。對應 TestComplete checkout-total。"""
        return self._js_text("checkout-total")

    def click_close_modal(self) -> None:
        """點擊關閉結帳 Modal 按鈕。對應 TestComplete close-modal。"""
        self._js_click("close-modal")

File: demo1_shoppingcart_website/Automation_testcase/Project_FlaUIBDD/Testcase_shopping_cart_FlaUI_BDD/PageObjects/ShoppingCartPage.cs
using FlaUI.Core.AutomationElements;
using FlaUI.Core.Definitions;
using FlaUI.Core.Exceptions;
using FlaUI.Core.Tools;
using FlaUI.UIA3;

namespace ShoppingCartTests.PageObjects
{
    /// <summary>
    /// 購物車頁面物件
    /// </summary>
    public class ShoppingCartPage : BasePage
    {
        // 元素定位器 (對應 HTML id，WebView2 會映射為 AutomationId)
        private const string CartCountId = "cart-count";
        private const string CartTotalId = "cart-total";
        private const string CartEmptyMessageId = "cart-empty-message";
        private const string ClearCartButtonId = "btn-clear-cart";
        private const string CheckoutButtonId = "btn-checkout";
        private const string CheckoutMessageId = "checkout-message";
        private const string CheckoutTotalId = "checkout-total";
        private const string CloseModalButtonId = "btn-close-modal";

        private const string QtyValuePrefix = "qty-value-";
        private const string QtyPlusPrefix = "btn-qty-plus-";
        private const string QtyMinusPrefix = "btn-qty-minus-";
        private const string RemoveItemPrefix = "btn-remove-";

        public ShoppingCartPage(Window window, UIA3Automation automation)
            : base(window, automation)
        {
        }

        #region 購物車基本操作

        /// <summary>
        /// 取得購物車件數
        /// </summary>
        public string GetCartCount()
        {
            return GetCartCountText();
        }

        /// <summary>
        /// 取得購物車總計
        /// </summary>
        public string GetCartTotal()
        {
            var count = GetCartCount();
            if (count == "0 件")
            {
                return "NT$ 0";
            }

            if (TryGetElementText(CartTotalId, out var total) && !string.IsNullOrWhiteSpace(total))
            {
                return total;
            }

            return GetLastTextMatchingPattern(@"^NT\$ \d+$");
        }

        /// <summary>
        /// 取得空購物車訊息
        /// </summary>
        public string GetCartEmptyMessage()
        {
            if (TryGetElementText(CartEmptyMessageId, out var message))
            {
                return message;
            }

            return GetTextMatchingPattern(@"^購物車是空的");
        }

        /// <summary>
        /// 點擊清空購物車按鈕
        /// </summary>
        public void ClickClearCart()
        {
            ClickElement(ClearCartButtonId);
        }

        /// <summary>
        /// 重置購物車（如果有商品則清空）
        /// </summary>
        public void ResetCart()
        {
            var count = GetCartCount();
            if (!string.IsNullOrEmpty(count) && count != "0 件")
            {
                ClickClearCart();
                WaitForWindow();
            }
        }

        #endregion

        #region 商品數量操作

        /// <summary>
        /// 取得指定商品的數量
        /// </summary>
        public string GetItemQuantity(string itemName)
        {
            var productId = ResolveProductId(itemName);
            var automationId = $"{QtyValuePrefix}{productId}";

            if (TryGetElementText(automationId, out var quantity, 1000) && !string.IsNullOrWhiteSpace(quantity))
            {
                return quantity;
            }

            var found = Retry.WhileNull(
                () => FindQuantityFromQtyControls(productId) ?? FindQuantityFromCartItem(productId),
                TimeSpan.FromSeconds(8)
            ).Result;

            if (found != null)
            {
                return found;
            }

            throw new ElementNotAvailableException($"找不到商品數量: {itemName}");
        }

        private string FindQuantityFromQtyControls(string productId)
        {
            var plusButtonId = $"{QtyPlusPrefix}{productId}";
            var minusButtonId = $"{QtyMinusPrefix}{productId}";
            var anchor = FindElementInRoots(plusButtonId) ?? FindElementInRoots(minusButtonId);

            if (anchor?.Parent == null)
            {
                return null;
            }

            return FindDigitTextInContainer(anchor.Parent);
        }

        private string FindQuantityFromCartItem(string productId)
        {
            var cartItem = FindElementInRoots($"cart-item-{productId}");
            if (cartItem == null)
            {
                return null;
            }

            return FindDigitTextInContainer(cartItem);
        }

        private static string FindDigitTextInContainer(AutomationElement container)
        {
            foreach (var child in container.FindAllDescendants())
            {
                if (child.ControlType == ControlType.Button)
                {
                    continue;
                }

                var text = ReadElementText(child);
                if (IsDigitsOnly(text))
                {
                    return text;
                }
            }

            return null;
        }

        /// <summary>
        /// 點擊增加商品數量按鈕
        /// </summary>
        public void ClickIncreaseQuantity(string itemName)
        {
            var automationId = $"{QtyPlusPrefix}{ResolveProductId(itemName)}";
            ClickElement(automationId);
        }

        /// <summary>
        /// 點擊減少商品數量按鈕
        /// </summary>
        public void ClickDecreaseQuantity(string itemName)
        {
            var automationId = $"{QtyMinusPrefix}{ResolveProductId(itemName)}";
            ClickElement(automationId);
        }

        /// <summary>
        /// 點擊移除商品按鈕
        /// </summary>
        public void ClickRemoveItem(string itemName)
        {
            var automationId = $"{RemoveItemPrefix}{ResolveProductId(itemName)}";
            ClickElement(automationId);
        }

        #endregion

        #region 結帳操作

        /// <summary>
        /// 點擊結帳按鈕
        /// </summary>
        public void ClickCheckout()
        {
            ClickElement(CheckoutButtonId);
        }

        /// <summary>
        /// 取得結帳訊息
        /// </summary>
        public string GetCheckoutMessage()
        {
            if (TryGetElementText(CheckoutMessageId, out var message, 3000) && !string.IsNullOrWhiteSpace(message))
            {
                return message;
            }

            return GetTextMatchingPattern(@"^感謝您的購買");
        }

        /// <summary>
        /// 取得結帳總金額
        /// </summary>
        public string GetCheckoutTotal()
        {
            if (TryGetElementText(CheckoutTotalId, out var total, 3000) && !string.IsNullOrWhiteSpace(total))
            {
                return total;
            }

            return GetTextMatchingPattern(@"^總金額：NT\$ \d+$");
        }

        /// <summary>
        /// 關閉結帳視窗
        /// </summary>
        public void CloseCheckoutModal()
        {
            ClickElement(CloseModalButtonId);
        }

        #endregion

        #region 驗證方法

        /// <summary>
        /// 驗證購物車件數
        /// </summary>
        public bool VerifyCartCount(string expectedCount)
        {
            return VerifyElementText(CartCountId, expectedCount);
        }

        /// <summary>
        /// 驗證購物車總計
        /// </summary>
        public bool VerifyCartTotal(string expectedTotal)
        {
            return VerifyElementText(CartTotalId, expectedTotal);
        }

        /// <summary>
        /// 驗證商品數量
        /// </summary>
        public bool VerifyItemQuantity(string itemName, string expectedQuantity)
        {
            var automationId = $"{QtyValuePrefix}{ResolveProductId(itemName)}";
            return VerifyElementText(automationId, expectedQuantity);
        }

        /// <summary>
        /// 驗證空購物車訊息
        /// </summary>
        public bool VerifyCartEmptyMessage(string expectedMessage)
        {
            return VerifyElementText(CartEmptyMessageId, expectedMessage);
        }

        /// <summary>
        /// 驗證結帳訊息
        /// </summary>
        public bool VerifyCheckoutMessage(string expectedMessage)
        {
            return VerifyElementText(CheckoutMessageId, expectedMessage);
        }

        /// <summary>
        /// 驗證結帳總金額
        /// </summary>
        public bool VerifyCheckoutTotal(string expectedTotal)
        {
            return VerifyElementText(CheckoutTotalId, expectedTotal);
        }

        #endregion

        private static string ResolveProductId(string itemName)
        {
            return itemName switch
            {
                "蘋果" => "apple",
                "香蕉" => "banana",
                "牛奶" => "milk",
                _ => itemName.ToLower()
            };
        }
    }
}
